### ISSUE TITLE
Regarding the Use by Chinese Users

### ISSUE THREAD

**Issue by ruixuan658** [vkbo/novelWriter#1657]
novelWriter is very user-friendly and also has some parts translated into Chinese. I recently discovered this project and liked it very much, and I contributed some Chinese translations. However, there is an issue for Chinese users:

Chinese, unlike English and other alphabetic languages, does not have the concept of letters, and the smallest unit for word count in Chinese is a character. In the current version of novelWriter, the word count includes 'characters' and 'words'. In current usage, for counting words in Chinese, one needs to look at 'characters', as 'words' does not make sense for counting Chinese characters. Additionally, in the current counting feature, 'words' count as a sentence in Chinese. I hope novelWriter can enhance the experience for Chinese users by improving this aspect.

**Comment by vkbo** [comment]
Incidentally, I'm redesigning the Project Details tool as well, where I want to add some more text analysis features. I know CJK character count has been requested before, and it's one of the additions I have in mind.

Basically, I want to define a tool box of text analysis and stats tools that the user can select from to generate a report for a given novel folder. This is something where user contributions may be very helpful. I will create a framework for it, and adding language-specific ones shouldn't be a problem.

**Comment by vkbo** [comment]
> novelWriter is very user-friendly and also has some parts translated into Chinese. I recently discovered this project and liked it very much, and I contributed some Chinese translations. 

On a side note, if you have the time to update the remaining missing translations (6% of the text), I can include a complete translation in the 2.2.1 release I'm planning really soon. Only complete translations are updated.

**Comment by ruixuan658** [comment]
Yes, it might be a good solution to set different character count rules for different languages in the preferences. Besides, during my use in these past few days, I have also found some Chinese translations that do not fit well in the Chinese context. I hope I can re-translate all the content into Chinese again.

**Comment by ruixuan658** [comment]
I don't know if you are familiar with languages like Chinese and Japanese, but I can give you a simple example:

> 这是一个很好的软件，我很喜欢。

This sentence means: "This is a very good software, I like it very much."

In the common Chinese character count, this sentence contains 13 characters.

In English, each word is separated by a space, so when using a character count feature not suitable for Chinese, this sentence would show as having 2 words. However, it has 13 characters. And in Chinese, the count of words is usually not considered. Perhaps in Chinese and similar languages, character counting doesn’t need to count words, but just the number of characters. In the current novelWriter, the 'characters' count is actually the number of Chinese characters, but the count for words is not the number of vocabulary words, but the number of sentences.

For example,

- “喜欢 (like)” is one Chinese word, but it contains two characters,
- “这 (this)” is also one Chinese word, but it contains one character.
- “我很喜欢 (I like it very much)” is a Chinese sentence, and not a single “word.”

**Comment by vkbo** [comment]
> Yes, it might be a good solution to set different character count rules for different languages in the preferences. Besides, during my use in these past few days, I have also found some Chinese translations that do not fit well in the Chinese context. I hope I can re-translate all the content into Chinese again.

Currently, @longqzh is has access to approve Chinese translations. @hebekeg is in charge of the Japanese translation.

**Comment by ruixuan658** [comment]
For Chinese, there are a few concepts that might need some explanation:
- 'Word Count' (字数): The number of Chinese characters.
- 'Character Count' (字符数): The total number of characters, including Chinese characters and symbols.

Regarding your questions:
- Japanese and Korean, like Chinese, are also based on character count for word statistics. However, I'm not particularly familiar with the specific word counting rules in Japanese since it uses both Kana and Kanji. Native speakers of Japanese and Korean might offer better suggestions on this matter.
- As a native Chinese speaker, in all the editors I've used, the primary metric for Chinese text has been 'Word Count' (字数), with 'Character Count' (字符数) as a secondary metric. The current paragraph and line count in novelWriter could perhaps be an added feature.
- Displaying 'Word Count' as 'Sentence Count' might not be the best approach. Strictly speaking, in Chinese, a sentence is primarily defined by the ending punctuation marks like periods and question marks, with commas (，) dividing what might be considered half a sentence.
- There's no issue with the way paragraph count is measured, using double line breaks is a good method. After all, in Markdown syntax, a single line break is a line break, and a double line break creates a new paragraph.
- In Microsoft Word, each Chinese character and each symbol are counted as one character. 'Character Count（字符数）' might also be a metric to consider — the total number of Chinese characters and symbols. In the panel below the project tree, displaying word count, character count, and paragraphs might be a good approach.
- In the project tree, it would be better to display word count, character count, and paragraphs, while in the status bar and writing statistics, showing the word count would be more appropriate.

**Comment by ruixuan658** [comment]
If the user is involved in submitting their work after completion, different submission platforms have different counting standards. Some platforms count the word count (only Chinese characters), while others count the character count (including Chinese characters and symbols). Therefore, for Chinese users, 'Word Count' and 'Character Count' might be the primary indicators for text statistics.

**Comment by ruixuan658** [comment]
The regular expression for pure Chinese characters (excluding symbols) is as follows: [\u4e00-\u9fa5].

**Comment by ruixuan658** [comment]
In the course of using it, I encountered another issue, which is that the Chinese input method has a candidate box. In NovelWriter, this candidate box obscures the existing input content, like this:
![image](https://github.com/vkbo/novelWriter/assets/60544711/7df2e757-0675-4fcd-b1ef-768569d38bff)
The correct display should be like this (as in Microsoft Word):
![image](https://github.com/vkbo/novelWriter/assets/60544711/d9e8df47-fae0-4747-ba99-6ca3a711c006)

**Comment by vkbo** [comment]
> In the course of using it, I encountered another issue, which is that the Chinese input method has a candidate box. In NovelWriter, this candidate box obscures the existing input content

There is no code in novelWriter to handle this. It likely comes from the Qt library or the system itself. What operating system are you using?

**Comment by vkbo** [comment]
>     * 'Word Count' (字数): The number of Chinese characters.

Is this what's sometimes called CJK count?

**Comment by ruixuan658** [comment]
> > In the course of using it, I encountered another issue, which is that the Chinese input method has a candidate box. In NovelWriter, this candidate box obscures the existing input content
> 
> There is no code in novelWriter to handle this. It likely comes from the Qt library or the system itself. What operating system are you using?

windows10，

**Comment by ruixuan658** [comment]
> > 'Word Count' (字数): The number of Chinese characters. 
> Is this what's sometimes called CJK count?

It should be, now in novelWriter, 'characters' is counted as 'Character Count' (字符数)，
>  The total number of characters, including Chinese characters and symbols.”

**Comment by vkbo** [comment]
> > > In the course of using it, I encountered another issue, which is that the Chinese input method has a candidate box. In NovelWriter, this candidate box obscures the existing input content
> > 
> > 
> > There is no code in novelWriter to handle this. It likely comes from the Qt library or the system itself. What operating system are you using?
> 
> windows10，

I plan to switch novelWriter to Qt6 some time this year. Probably in release 2.4. Hopefully that will help, but a bit of searching reveals the positioning of the candidate box is a problem for many apps. This requires a bit of research. I suggest you make a bug report from your comment so we can track it separately.

**Comment by ruixuan658** [comment]
I consulted chatGPT, and it might be an issue with the QT5 package. Here is their response:

> I found information that suggests issues with the position of the input method candidate window in Qt applications are not uncommon. These issues seem to occur across different platforms and environments, and can be related to several factors including high DPI settings, the version of the underlying system libraries, or specific configurations in the Qt environment.
> 
> One issue that was reported involved the candidate window appearing at an incorrect position in relation to the input control, particularly in high DPI settings. The problem seemed to resolve when the scale setting was changed to 100%, indicating that scaling factors can influence the candidate window's position. This issue was not specific to Electron-based applications but was observed in various environments including Ubuntu and Fedora, under different desktop environments like GNOME and KDE.
> 
> Another reported issue was with the IBus input method in a Qt-based terminal widget. The candidate window position was not updating correctly due to the absence of specific code that would update the IM's cursor position. The problem was resolved by adding code to ensure the IM's cursor position was updated appropriately when the output changed.
> 
> These cases suggest that the issue you're experiencing could be related to the way the Qt environment interacts with the input method framework and the specific configurations or scaling settings of your system. It might be helpful to check the scaling settings, look into how the input method's cursor position is being managed in the application, and ensure all relevant libraries and frameworks are up to date. Additionally, considering the input method's compatibility with the version of Qt being used could provide some insights.

**Comment by vkbo** [comment]
ChatGPT is not a reliable source. It is mixing up a bunch of different issues in that answer too. I'm familiar with the Linux issues with positioning. I will do some proper research.

**Comment by ruixuan658** [comment]
> ChatGPT is not a reliable source. It is mixing up a bunch of different issues in that answer too. I'm familiar with the Linux issues with positioning. I will do some proper research.

Thank you very much, I am not sure if this issue exists in the Linux system, as I am using the Windows 10 operating system.

**Comment by vkbo** [comment]
The Linux issue mostly has to do with determining the (0,0) coordinate of the screen on multi-screen setups where the monitors have different resolution or pixel scaling. AFAIK this is related to Wayland. This is much better supported in Qt6.

On Windows, I'm not sure what the issue is. But in general, the way Qt computes coordinates for pop up boxes, including menus, depend on getting information from the window manager in the operating system. The Qt5 library is not up to date with various such features, including high DPI scaling, light/dark OS settings, etc. A lot of this has been fixed in Qt6.

Anyway, please make a proper bug ticket on this so I can track it as a bug.

**Comment by longqzh** [comment]
@ruixuan658 , @vkbo 

As a native Chiense and Korean speaker, first of all I'd like to redefine some conceptions in order to avoid confusion.
- **Conception A** : character count with blank
  - the sum of the number of characters and symbols. it excludes blank number
- **Conception B** : character count without blank
  - the sum of the number of characters, symbols and blanks.
- **Conception C** : word count 
  - the block number between blanks

> Can we treat Chinese, Japanese and Korean in the same way here?

**No** , because in Korean, there are blank between words, but in Chinese, there isn't any blank between words. 
And if you want to split the words in a sentence properly, you have to use some professional library like [jieba](https://github.com/fxsjy/jieba)

So **Conception C** makes no sense in Chinese. And currently `Word Count` in novelwriter means sentence number because there is only blank between Chinese sentence : )

> Character count is clearly the primary metric. Is the current way characters are counted also usable for CJK, or do we need a different method?

For Chinese, **Conception A** and **Conception B** is enough. For Korean, **Conception B** is also necessary. I don't think we need other new method.

> Can we display the current "word count" as a sentence count? Is that a useful parameter?

If we don't change the algorithm about "word count", in Chinese it means sentence count naturally. But I don't think it's a useful parameter, I have never seen it in others normal editor ( I'm not profesional editer's user )

> Is the paragraph count still valid? It counts double line breaks.

I'm not sure. To be honest, maybe we should force the user to use double line breaks between paragraphes.


> What three counts can we show in the panel below the project tree where it now shows characters, words and paragraphs?

`characters` and `paragraphs` should be shown in panel, if possible maybe we should distinguish **Conception A** and **Conception B** in `charaters.  
And `words` isn't easy to calcualte in Chinese. I don't think we should show it.

**Comment by vkbo** [comment]
> > Can we treat Chinese, Japanese and Korean in the same way here?
> 
> **No** , because in Korean, there are blank between words, but in Chinese, there isn't any blank between words. By the way, it's **VERY HARD** to get the word count in Chinese. If you want to split the Chinese words in a sentence properly, you have to use some professional library like [jieba](https://github.com/fxsjy/jieba).

I really want to avoid adding an external library here. Both because I like to avoid dependencies whenever possible, and because the word count algorithm is performance critical, and I need to have control over how it's optimised.

So my real question is, will a CJK character count suffice? It's a fairly easy counter to implement.

> For Chinese, **Conception A** and **Conception B** is enough. For Korean, **Conception C** is also necessary. I don't think we need other new method.

How necessary? The CJK count will ignore spaces. I propose to use CJK count as a drop-in replacement for word count, and just remove word count when it is enabled.

Basically I would like to add a new setting in Preferences called "Primary Counter" or something, with the following options:

1. Latin Word Count
2. CJK Character Count

> Generally speaking, I prefer to just show `characters count` in Chinese and hide the `word count`. For Korean and Japanse, treat them as English.

With an option like above, the user could select their preferred counter. I'm also working on adding text analysis tools which will include a bunch of different counter methods, so CJK count will always be available, but this is mostly about which metric is used for collecting statistics and displaying in the project tree.

**Comment by longqzh** [comment]
@vkbo 

> So my real question is, will a CJK character count suffice? It's a fairly easy counter to implement.

For Chinese and Korean, character count is enough. I also think word couting is pointless, especially given the complexity of the implementation :) 

> How necessary? The CJK count will ignore spaces. I propose to use CJK count as a drop-in replacement for word count, and just remove word count when it is enabled.

I agree with you. character count is much more importance than word count.

**Comment by vkbo** [comment]
Just to clarify, CJK character count is not the same as the current character count.

**Comment by longqzh** [comment]
> We also need to discuss how such a counter is implemented. Python uses utf-8, so the absolute simplest option is to count Unicode values between \u4e00 and \u9fff. I'm not sure if that is sufficient, or more ranges should be included?

To clarify Python use **Unicode** to represent all character in the python virtual machine [Official Docs](https://docs.python.org/3/howto/unicode.html). 
When we try to output the text (save to hard disk, send to network, etc), we need to **encode** the text to a given specification (ascii, utf-8, unicode, ...)
So please recheck it 😄

**Comment by vkbo** [comment]
> So please recheck it 😄

Python 3 strings are utf-8 always. This is what the count is run on. This is not an issue.

The question is what Unicode ranges we need to check for CJK count.

**Comment by vkbo** [comment]
In any case, I have full control over the encoding in novelWriter. Both novelWriter and the Qt framework use Unicode (utf-8 and utf-16 respectively). What I need input on is what Unicode ranges will cover the correct symbols needed to make a proper count. I don't know the differences between the ones listed on that wiki page.

**Comment by deyigifts** [comment]
The Conception B looks promising, since word split of Chinese sentence usually reply on the natural language processing(NLP)  module.
Since the post metadata got the language type, it should be possible to add a switch for word count implements with CJK contents.

**Comment by vkbo** [comment]
Concept B has been available for the Manuscript tool for some time now. But it is not implemented in project view as an alternative to word counts.

I am still waiting for feedback on whether CJK count is desirable, based on Unicode ranges.

**Comment by ZEROIC** [comment]
As my former tool Obsidian,the famous plugin "[better word count](https://github.com/lukeleppan/better-word-count)" implements like this:

```typescript
export function getWordCount(text: string): number {
  const spaceDelimitedChars =
    /'’A-Za-z\u00AA\u00B5\u00BA\u00C0-\u00D6\u00D8-\u00F6\u00F8-\u02C1\u02C6-\u02D1\u02E0-\u02E4\u02EC\u02EE\u0370-\u0374\u0376\u0377\u037A-\u037D\u037F\u0386\u0388-\u038A\u038C\u038E-\u03A1\u03A3-\u03F5\u03F7-\u0481\u048A-\u052F\u0531-\u0556\u0559\u0561-\u0587\u05D0-\u05EA\u05F0-\u05F2\u0620-\u064A\u066E\u066F\u0671-\u06D3\u06D5\u06E5\u06E6\u06EE\u06EF\u06FA-\u06FC\u06FF\u0710\u0712-\u072F\u074D-\u07A5\u07B1\u07CA-\u07EA\u07F4\u07F5\u07FA\u0800-\u0815\u081A\u0824\u0828\u0840-\u0858\u08A0-\u08B4\u0904-\u0939\u093D\u0950\u0958-\u0961\u0971-\u0980\u0985-\u098C\u098F\u0990\u0993-\u09A8\u09AA-\u09B0\u09B2\u09B6-\u09B9\u09BD\u09CE\u09DC\u09DD\u09DF-\u09E1\u09F0\u09F1\u0A05-\u0A0A\u0A0F\u0A10\u0A13-\u0A28\u0A2A-\u0A30\u0A32\u0A33\u0A35\u0A36\u0A38\u0A39\u0A59-\u0A5C\u0A5E\u0A72-\u0A74\u0A85-\u0A8D\u0A8F-\u0A91\u0A93-\u0AA8\u0AAA-\u0AB0\u0AB2\u0AB3\u0AB5-\u0AB9\u0ABD\u0AD0\u0AE0\u0AE1\u0AF9\u0B05-\u0B0C\u0B0F\u0B10\u0B13-\u0B28\u0B2A-\u0B30\u0B32\u0B33\u0B35-\u0B39\u0B3D\u0B5C\u0B5D\u0B5F-\u0B61\u0B71\u0B83\u0B85-\u0B8A\u0B8E-\u0B90\u0B92-\u0B95\u0B99\u0B9A\u0B9C\u0B9E\u0B9F\u0BA3\u0BA4\u0BA8-\u0BAA\u0BAE-\u0BB9\u0BD0\u0C05-\u0C0C\u0C0E-\u0C10\u0C12-\u0C28\u0C2A-\u0C39\u0C3D\u0C58-\u0C5A\u0C60\u0C61\u0C85-\u0C8C\u0C8E-\u0C90\u0C92-\u0CA8\u0CAA-\u0CB3\u0CB5-\u0CB9\u0CBD\u0CDE\u0CE0\u0CE1\u0CF1\u0CF2\u0D05-\u0D0C\u0D0E-\u0D10\u0D12-\u0D3A\u0D3D\u0D4E\u0D5F-\u0D61\u0D7A-\u0D7F\u0D85-\u0D96\u0D9A-\u0DB1\u0DB3-\u0DBB\u0DBD\u0DC0-\u0DC6\u0E01-\u0E30\u0E32\u0E33\u0E40-\u0E46\u0E81\u0E82\u0E84\u0E87\u0E88\u0E8A\u0E8D\u0E94-\u0E97\u0E99-\u0E9F\u0EA1-\u0EA3\u0EA5\u0EA7\u0EAA\u0EAB\u0EAD-\u0EB0\u0EB2\u0EB3\u0EBD\u0EC0-\u0EC4\u0EC6\u0EDC-\u0EDF\u0F00\u0F40-\u0F47\u0F49-\u0F6C\u0F88-\u0F8C\u1000-\u102A\u103F\u1050-\u1055\u105A-\u105D\u1061\u1065\u1066\u106E-\u1070\u1075-\u1081\u108E\u10A0-\u10C5\u10C7\u10CD\u10D0-\u10FA\u10FC-\u1248\u124A-\u124D\u1250-\u1256\u1258\u125A-\u125D\u1260-\u1288\u128A-\u128D\u1290-\u12B0\u12B2-\u12B5\u12B8-\u12BE\u12C0\u12C2-\u12C5\u12C8-\u12D6\u12D8-\u1310\u1312-\u1315\u1318-\u135A\u1380-\u138F\u13A0-\u13F5\u13F8-\u13FD\u1401-\u166C\u166F-\u167F\u1681-\u169A\u16A0-\u16EA\u16F1-\u16F8\u1700-\u170C\u170E-\u1711\u1720-\u1731\u1740-\u1751\u1760-\u176C\u176E-\u1770\u1780-\u17B3\u17D7\u17DC\u1820-\u1877\u1880-\u18A8\u18AA\u18B0-\u18F5\u1900-\u191E\u1950-\u196D\u1970-\u1974\u1980-\u19AB\u19B0-\u19C9\u1A00-\u1A16\u1A20-\u1A54\u1AA7\u1B05-\u1B33\u1B45-\u1B4B\u1B83-\u1BA0\u1BAE\u1BAF\u1BBA-\u1BE5\u1C00-\u1C23\u1C4D-\u1C4F\u1C5A-\u1C7D\u1CE9-\u1CEC\u1CEE-\u1CF1\u1CF5\u1CF6\u1D00-\u1DBF\u1E00-\u1F15\u1F18-\u1F1D\u1F20-\u1F45\u1F48-\u1F4D\u1F50-\u1F57\u1F59\u1F5B\u1F5D\u1F5F-\u1F7D\u1F80-\u1FB4\u1FB6-\u1FBC\u1FBE\u1FC2-\u1FC4\u1FC6-\u1FCC\u1FD0-\u1FD3\u1FD6-\u1FDB\u1FE0-\u1FEC\u1FF2-\u1FF4\u1FF6-\u1FFC\u2071\u207F\u2090-\u209C\u2102\u2107\u210A-\u2113\u2115\u2119-\u211D\u2124\u2126\u2128\u212A-\u212D\u212F-\u2139\u213C-\u213F\u2145-\u2149\u214E\u2183\u2184\u2C00-\u2C2E\u2C30-\u2C5E\u2C60-\u2CE4\u2CEB-\u2CEE\u2CF2\u2CF3\u2D00-\u2D25\u2D27\u2D2D\u2D30-\u2D67\u2D6F\u2D80-\u2D96\u2DA0-\u2DA6\u2DA8-\u2DAE\u2DB0-\u2DB6\u2DB8-\u2DBE\u2DC0-\u2DC6\u2DC8-\u2DCE\u2DD0-\u2DD6\u2DD8-\u2DDE\u2E2F\u3005\u3006\u3031-\u3035\u303B\u303C\u3105-\u312D\u3131-\u318E\u31A0-\u31BA\u31F0-\u31FF\u3400-\u4DB5\uA000-\uA48C\uA4D0-\uA4FD\uA500-\uA60C\uA610-\uA61F\uA62A\uA62B\uA640-\uA66E\uA67F-\uA69D\uA6A0-\uA6E5\uA717-\uA71F\uA722-\uA788\uA78B-\uA7AD\uA7B0-\uA7B7\uA7F7-\uA801\uA803-\uA805\uA807-\uA80A\uA80C-\uA822\uA840-\uA873\uA882-\uA8B3\uA8F2-\uA8F7\uA8FB\uA8FD\uA90A-\uA925\uA930-\uA946\uA960-\uA97C\uA984-\uA9B2\uA9CF\uA9E0-\uA9E4\uA9E6-\uA9EF\uA9FA-\uA9FE\uAA00-\uAA28\uAA40-\uAA42\uAA44-\uAA4B\uAA60-\uAA76\uAA7A\uAA7E-\uAAAF\uAAB1\uAAB5\uAAB6\uAAB9-\uAABD\uAAC0\uAAC2\uAADB-\uAADD\uAAE0-\uAAEA\uAAF2-\uAAF4\uAB01-\uAB06\uAB09-\uAB0E\uAB11-\uAB16\uAB20-\uAB26\uAB28-\uAB2E\uAB30-\uAB5A\uAB5C-\uAB65\uAB70-\uABE2\uAC00-\uD7A3\uD7B0-\uD7C6\uD7CB-\uD7FB\uF900-\uFA6D\uFA70-\uFAD9\uFB00-\uFB06\uFB13-\uFB17\uFB1D\uFB1F-\uFB28\uFB2A-\uFB36\uFB38-\uFB3C\uFB3E\uFB40\uFB41\uFB43\uFB44\uFB46-\uFBB1\uFBD3-\uFD3D\uFD50-\uFD8F\uFD92-\uFDC7\uFDF0-\uFDFB\uFE70-\uFE74\uFE76-\uFEFC\uFF21-\uFF3A\uFF41-\uFF5A\uFF66-\uFFBE\uFFC2-\uFFC7\uFFCA-\uFFCF\uFFD2-\uFFD7\uFFDA-\uFFDC/
      .source;
  const nonSpaceDelimitedWords =
    /\u3041-\u3096\u309D-\u309F\u30A1-\u30FA\u30FC-\u30FF\u4E00-\u9FD5/.source;

  const nonSpaceDelimitedWordsOther =
    /[\u3041-\u3096\u309D-\u309F\u30A1-\u30FA\u30FC-\u30FF\u4E00-\u9FD5]{1}/
      .source;

  const pattern = new RegExp(
    [
      `(?:[0-9]+(?:(?:,|\\.)[0-9]+)*|[\\-${spaceDelimitedChars}])+`,
      nonSpaceDelimitedWords,
      nonSpaceDelimitedWordsOther,
    ].join("|"),
    "g"
  );
  return (text.match(pattern) || []).length;
}

export function getCharacterCount(text: string): number {
  return text.length;
}
```
By the way,I usually focus total character counts more,because all none-blank characters(punctuation marks included or not,depends on rules of the editor) are counted if I submit a chinese article.Counting exact "chinese words" is unnecessary.

**Comment by vkbo** [comment]
Thanks! This is useful, and information I've been waiting for to complete this.

**Comment by vkbo** [comment]
In 2.7 a new setting is added to Preferences where character counts can be set as the preferred number to display on the UI. This preference will only be applied to the project and novel tree for now, and I'll update the editor footer and status bar next.

Also, the character count change between sessions will now also be tracked in the session log, so statistics on change of character count will begin to accumulate. I intend to update the writing statistics tool at some point, so I may delay adding character support there until then.
